# 1. Orchestration 

#### Complete the setup of a swarm mode cluster, with managers and worker nodes

In [6]:
%%bash 
docker swarm init

Swarm initialized: current node (jp3yeaijq1ne69evozl95hfc1) is now a manager.

To add a worker to this swarm, run the following command:

    docker swarm join --token SWMTKN-1-5tmjlpf3obk5owtpvgpuwbo4vvjjhjiqi7lng9gz2sthci2kdg-3ubvdmj0eu9drvpg9vjavm6n7 10.10.1.60:2377

To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.



To join additional nodes to the swarm

    docker swarm join --token SWMTKN-1-5tmjlpf3obk5owtpvgpuwbo4vvjjhjiqi7lng9gz2sthci2kdg-3ubvdmj0eu9drvpg9vjavm6n7 10.10.1.60:2377

Get the join command again

    docker swarm join-token worker 
    docker swarm join-token manager


#### State the differences between running a container vs running a service

Containers can be managed in different ways and depends on the application. You can run containers and manage them individually by using 'docker run' or when running in Swarm mode ability to use 'docker service'. Docker service allows running multiple containers and defining how they should be spread across nodes for availability, deployment stratergies, the network type, port etc..

docker service is the new docker run 

http://events.linuxfoundation.org/sites/events/files/slides/ContainerCon%20Berlin%20%28Goelzer%29%20-%20Upload%209-18-2016.pdf

#### Demonstrate steps to lock a swarm cluster

Auto lock the swarm

    docker swarm init --auto-lock 
    docker swarm update --autolock=true


#### Extend the instructions to run individual containers into running services under swarm

docker service create --name testweb -p 80:80 httpd

In [10]:
%%bash
docker service ls

ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
pq1i6airnwry        testweb             replicated          1/1                 httpd:latest        *:80->80/tcp


#### Interpret the output of "docker inspect" commands

In [2]:
%%bash 
docker inspect centos

[
    {
        "Id": "sha256:1e1148e4cc2c148c6890a18e3b2d2dde41a6745ceb4e5fe94a923d811bf82ddb",
        "RepoTags": [
            "centos:latest"
        ],
        "RepoDigests": [
            "centos@sha256:184e5f35598e333bfa7de10d8fb1cebb5ee4df5bc0f970bf2b1e7c7345136426"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "2018-12-06T00:21:07.135655444Z",
        "Container": "1fdbb0fcc184eb795364f7aa5fdc00299d0a2b90d8e26b4696217c22da7f983f",
        "ContainerConfig": {
            "Hostname": "1fdbb0fcc184",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(

ID: When running 'docker images' this is the Image ID, a sha256 digest of the JSON config.  
RepoTags: The version of the docker image 
RepoDigests:
Parent: The container this image was based off
Comment: Comments within the dockerfile 
Created: The time the image was created 
Container: 
ContainerConfig: Container this image was created from
 Hostname: Name of the container
 Domainname: Domainname for container 
 User: The user this container should use 
 AttachStdin: Attach local standard input stream into a running container
 Tty: Text input output environment 
 OpenStdin:
 StdinOnce:
 Env: Environmemnt variable 
 Cmd: Commands being run 
 ArgsEscaped: 
 Image: Image used within dockerfile
 Volumes: Persistent data 
 WorkingDir: Sets working directory for commands run on the dockerfile 
 Entrypoint: This might be something like /bin/bash or /bin/sh
 OnBuild: Executes after the dockerfile build completes. Executes in any child image derived from current image. 
 Labels: Key-Value pair of image information 
DockerVersion: Docker version information 
Author: This is taken from maintainer information 
Config:
Architecture: Architecture type 
Os: Operating System 
Size: Size of Image  
VirtualSize: Size of Image 
GraphDriver: 
RootFS
Metadata: 

#### Convert an application deployment into a stack file using a YAML compose file with "docker stack deploy"

In [12]:
%%bash 
sudo curl -L "https://github.com/docker/compose/releases/download/1.23.2/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   617    0   617    0     0   1615      0 --:--:-- --:--:-- --:--:--  1615
100 11.2M  100 11.2M    0     0  7201k      0  0:00:01  0:00:01 --:--:-- 20.3M


In [13]:
%%bash
sudo chmod +x /usr/local/bin/docker-compose

In [15]:
%%bash
sudo ln -s /usr/local/bin/docker-compose /usr/bin/docker-compose

In [18]:
%%bash
docker-compose --version

docker-compose version 1.23.2, build 1110ad01


In [2]:
%%bash
docker stack deploy --compose-file docker-compose.yml stack

Creating service stack_redis
Creating service stack_web


In [3]:
%%bash
docker service ls 

ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
1k0nmot8tr3s        stack_redis         replicated          1/1                 redis:latest        
d6zhy2je0j0p        stack_web           replicated          1/1                 httpd:latest        *:5000->5000/tcp


#### Manipulate a running stack of services

In [8]:
%%bash
docker service rm stack_redis 

stack_redis


In [19]:
%%bash
docker service update --limit-cpu=.5 --reserve-cpu=1 --limit-memory=128m --reserve-memory=256m centos

Error: No such service: centos


#### Increase # of replicas

In [9]:
%%bash 
docker service update --replicas 3 stack_web

stack_web
overall progress: 0 out of 3 tasks
1/3:  
2/3:  
3/3:  
overall progress: 3 out of 3 tasks
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that task

#### Add networks, publish ports

In [15]:
%%bash 
docker service update --publish-add 8080 stack_web

stack_web
overall progress: 0 out of 3 tasks
1/3:  
2/3:  
3/3:  
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 0 out of 3 tasks
overall progress: 1 out of 3 tasks
overall progress: 1 out of 3 tasks
overall progress: 1 out 

In [16]:
%%bash
docker service ls

ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
d6zhy2je0j0p        stack_web           replicated          3/3                 httpd:latest        *:5000->5000/tcp, *:30001->80/tcp, *:30002->8080/tcp


#### Mount volumes

In [23]:
%%bash 
docker volume create datavol

datavol


In [26]:
%%bash 
sudo cp -r *  /var/lib/docker/volumes/datavol/_data/

In [7]:
%%bash 
docker run -d --mount source=datavol,target=/app centos:latest

5d9d3cafe13cbfec7c32193a38ab9f2e386b07142e50e3ce7410a24c14cc93a2


#### Illustrate running a replicated vs global service

![title](replicated-vs-global.png)

Replicated service, you specify the number of identical tasks you want to run. For example, you decide to deploy an HTTP service with three replicas, each serving the same content.

Global service is a service that runs one task on every node. There is no pre-specified number of tasks. Each time you add a node to the swarm, the orchestrator creates a task and the scheduler assigns the task to the new node. Good candidates for global services are monitoring agents, an anti-virus scanners or other types of containers that you want to run on every node in the swarm.

In [5]:
%%bash
docker service create httpd

07ps2ju9h0qx0qirkuurnugy5
overall progress: 0 out of 1 tasks
1/1:  
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 1 out of 1 tasks
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 secon

In [7]:
%%bash
docker service ls 

ID                  NAME                    MODE                REPLICAS            IMAGE               PORTS
gkpb4ffb6vz0        condescending_khorana   replicated          0/1                 centos:latest       
07ps2ju9h0qx        determined_lamport      replicated          1/1                 httpd:latest        
yvibump8alf7        loving_hopper           replicated          0/3                 centos:latest       
c75dl6gb6qt1        nifty_kowalevski        replicated          0/1                 centos:latest       
d6zhy2je0j0p        stack_web               replicated          3/3                 httpd:latest        *:5000->5000/tcp, *:30001->80/tcp, *:30002->8080/tcp


In [8]:
%%bash
docker service update --replicas 3 determined_lamport

determined_lamport
overall progress: 0 out of 3 tasks
1/3:  
2/3:  
3/3:  
overall progress: 1 out of 3 tasks
overall progress: 1 out of 3 tasks
overall progress: 1 out of 3 tasks
overall progress: 1 out of 3 tasks
overall progress: 1 out of 3 tasks
overall progress: 3 out of 3 tasks
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tas

In [10]:
%%bash 
docker service ls | grep determined_lamport

07ps2ju9h0qx        determined_lamport      replicated          3/3                 httpd:latest        


In [13]:
%%bash 
docker service create -p 5901:80 --mode global httpd

f9we96x8zbl863v4326g0b3ct
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 0 out of 1 tasks
overall progress: 1 out of 1 tasks
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 5 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 4 seconds to verify that tasks are stable...
verify: Waiting 3 seconds to verify that tasks are stable...
verify: Waiting 3 secon

#### Identify the steps needed to troubleshoot a service not deploying

#### Apply node labels to demonstrate placement of tasks

In [ ]:
%%bash
docker service create --name constraints -p 80:80 --constraint 'node.labels.mynode == mainnode' --replicas 3 httpd

#### Sketch how a Dockerized application communicates with legacy systems

#### Paraphrase the importance of quorum in a swarm cluster

When the Docker Engine runs in swarm mode, manager nodes implement the Raft Consensus Algorithm to manage the global cluster state.

The reason why Docker swarm mode is using a consensus algorithm is to make sure that all the manager nodes that are in charge of managing and scheduling tasks in the cluster, are storing the same consistent state.

Having the same consistent state across the cluster means that in case of a failure, any Manager node can pick up the tasks and restore the services to a stable state. For example, if the Leader Manager which is responsible for scheduling tasks in the cluster dies unexpectedly, any other Manager can pick up the task of scheduling and re-balance tasks to match the desired state.

Systems using consensus algorithms to replicate logs in a distributed systems do require special care. They ensure that the cluster state stays consistent in the presence of failures by requiring a majority of nodes to agree on values.

Raft tolerates up to (N-1)/2 failures and requires a majority or quorum of (N/2)+1 members to agree on values proposed to the cluster. This means that in a cluster of 5 Managers running Raft, if 3 nodes are unavailable, the system cannot process any more requests to schedule additional tasks. The existing tasks keep running but the scheduler cannot rebalance tasks to cope with failures if the manager set is not healthy.

#### Demonstrate the usage of templates with "docker service create"

In [ ]:
%%bash
docker service create --name myweb --hostname="{{.Node.ID}}-{{.Service.Name}}" httpd5%